Failed to apply jb to some method.

Please examine each of the following points so that we can help you as soon and best as possible.

**Describe the bug**
While calling `retrieveActiveBody` on some method, soot will throws RuntimeException: Failed to apply jb to <org.springframework.http.client.Netty4ClientHttpRequest: org.springframework.util.concurrent.ListenableFuture executeInternal(org.springframework.http.HttpHeaders)>

**Input file**
[Spring-Web 5.2.2.jar](https://repo1.maven.org/maven2/org/springframework/spring-web/5.2.2.RELEASE/spring-web-5.2.2.RELEASE.jar)

**To reproduce**

Soot cofig option:
```scala
    Options.v().set_app(false)
    Options.v().set_debug(false)
    Options.v().set_verbose(false)
    Options.v().set_validate(false)
    Options.v().set_whole_program(true) // false got less phantom class
    Options.v().set_wrong_staticness(Options.wrong_staticness_ignore)
    // keep debugging info
    Options.v().set_keep_line_number(true)
    Options.v().set_keep_offset(true)
    // ignore library code
    Options.v().set_no_bodies_for_excluded(true)
    Options.v().set_allow_phantom_refs(true)
    // keep variable names
    Options.v().setPhaseOption("jb", "use-original-names:true")
    Options.v().setPhaseOption("jb.sils", "enabled:false")

    // Options.v().set_no_writeout_body_releasing(true);
    Options.v().set_drop_bodies_after_load(false);
```

Processing code:
```scala
    val total = Scene.v.getApplicationClasses.size()
    println(s"Start parsing $total classes")
    for ((c, i) <- Scene.v.getApplicationClasses().zipWithIndex) {
      for (m <- c.getMethods()) {
        if (!Modifier.isAbstract(m.getModifiers()) && !Modifier.isNative(m.getModifiers())) {
          try {
            val body = m.retrieveActiveBody()
          } catch {
            case ex: Exception => ex.printStackTrace()
          }
        }
      }
    }
```

**Expected behavior**
A clear and concise description of what you expected to happen.

**Stacktrace**

```
java.lang.RuntimeException: Failed to apply jb to <org.springframework.http.client.Netty4ClientHttpRequest: org.springframework.util.concurrent.ListenableFuture executeInternal(org.springframework.http.HttpHeaders)>
	at soot.asm.AsmMethodSource.getBody(AsmMethodSource.java:2281)
	at soot.SootMethod.retrieveActiveBody(SootMethod.java:447)
        ...

Caused by: java.lang.RuntimeException: This operation requires resolving level HIERARCHY but io.netty.channel.ChannelFutureListener is at resolving level DANGLING
If you are extending Soot, try to add the following call before calling soot.Main.main(..):
Scene.v().addBasicClass(io.netty.channel.ChannelFutureListener,HIERARCHY);
Otherwise, try whole-program mode (-w).
	at soot.SootClass.checkLevelIgnoreResolving(SootClass.java:210)
	at soot.SootClass.checkLevel(SootClass.java:192)
	at soot.FastHierarchy.canStoreClass(FastHierarchy.java:384)
	at soot.FastHierarchy.canStoreType(FastHierarchy.java:293)
	at soot.jimple.toolkits.typing.fast.BytecodeHierarchy.ancestor_(BytecodeHierarchy.java:230)
	at soot.jimple.toolkits.typing.fast.BytecodeHierarchy.ancestor(BytecodeHierarchy.java:288)
	at soot.jimple.toolkits.typing.fast.TypeResolver$CastInsertionUseVisitor.visit(TypeResolver.java:224)
	at soot.jimple.toolkits.typing.fast.IUseVisitor.visit(IUseVisitor.java:38)
	at soot.jimple.toolkits.typing.fast.UseChecker.handleInvokeExpr(UseChecker.java:153)
	at soot.jimple.toolkits.typing.fast.UseChecker.caseAssignStmt(UseChecker.java:416)
	at soot.jimple.internal.JAssignStmt.apply(JAssignStmt.java:217)
	at soot.jimple.toolkits.typing.fast.UseChecker.check(UseChecker.java:140)
	at soot.jimple.toolkits.typing.fast.TypeResolver.insertCasts(TypeResolver.java:369)
	at soot.jimple.toolkits.typing.fast.TypeResolver.minCasts(TypeResolver.java:392)
	at soot.jimple.toolkits.typing.fast.TypeResolver.inferTypes(TypeResolver.java:159)
	at soot.jimple.toolkits.typing.TypeAssigner.internalTransform(TypeAssigner.java:129)
	at soot.BodyTransformer.transform(BodyTransformer.java:47)
	at soot.Transform.apply(Transform.java:126)
	at soot.JimpleBodyPack.applyPhaseOptions(JimpleBodyPack.java:79)
	at soot.JimpleBodyPack.internalApply(JimpleBodyPack.java:112)
	at soot.Pack.apply(Pack.java:126)
	at soot.asm.AsmMethodSource.getBody(AsmMethodSource.java:2279)
	... 39 more
```


**Additional context**
I've also tried whole-program mode (-w) and got the same result.

Just for your information, the source code of the method which throws exception is:

https://github.com/spring-projects/spring-framework/blob/5.2.x/spring-web/src/main/java/org/springframework/http/client/Netty4ClientHttpRequest.java#L113

In the method a lambda experession is assigned to functional interface `io.netty.channel.ChannelFutureListener`, which is not in classpath so the resolving level is `DANGLING`. Is there any workaround to still get Jimple body of that method anyway? 

BTW, the bytecode of `ChannelFutureListener connectionListener = future -> {}` is:
```
invokedynamic java/lang/invoke/LambdaMetafactory.metafactory(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/MethodType;Ljava/lang/invoke/MethodType;Ljava/lang/invoke/MethodHandle;Ljava/lang/invoke/MethodType;)Ljava/lang/invoke/CallSite; : operationComplete(Lorg/springframework/http/client/Netty4ClientHttpRequest;Lorg/springframework/util/concurrent/SettableListenableFuture;Lorg/springframework/http/HttpHeaders;)Lio/netty/channel/ChannelFutureListener; (Lio/netty/util/concurrent/Future;)V org/springframework/http/client/Netty4ClientHttpRequest.lambda$executeInternal$0(Lorg/springframework/util/concurrent/SettableListenableFuture;Lorg/springframework/http/HttpHeaders;Lio/netty/channel/ChannelFuture;)V (7) (Lio/netty/channel/ChannelFuture;)V
```

In the messages it states 
```Caused by: java.lang.RuntimeException: This operation requires resolving level HIERARCHY but io.netty.channel.ChannelFutureListener is at resolving level DANGLING
If you are extending Soot, try to add the following call before calling soot.Main.main(..):
Scene.v().addBasicClass(io.netty.channel.ChannelFutureListener,HIERARCHY);
Otherwise, try whole-program mode (-w).
```

Adding `Scene.v().addBasicClass(io.netty.channel.ChannelFutureListener,HIERARCHY);` before you load classes should work.

In a case we've dealt with we added it before `Scene.v().loadNecessaryClasses();`
The lambda expression does not exist in the input so you cannot get the method body of the code.

--- a/src/main/java/soot/AbstractASMBackend.java
+++ b/src/main/java/soot/AbstractASMBackend.java
@@ -81,6 +81,7 @@ import soot.tagkit.SourceFileTag;
 import soot.tagkit.SyntheticTag;
 import soot.tagkit.Tag;
 import soot.tagkit.VisibilityAnnotationTag;
+import soot.tagkit.VisibilityLocalVariableAnnotationTag;
 import soot.tagkit.VisibilityParameterAnnotationTag;
 import soot.util.backend.ASMBackendUtils;
 import soot.util.backend.SootASMClassWriter;
@@ -394,7 +395,16 @@ public abstract class AbstractASMBackend {
       if (mv != null) {
         // Visit parameter annotations
         for (Tag t : sm.getTags()) {
-          if (t instanceof VisibilityParameterAnnotationTag) {
+          if (t instanceof VisibilityLocalVariableAnnotationTag) {
+            // Actually supporting this feature would require quite a bit of work.
+            // The problem is that we do not retain all relevant information yet;
+            // these annotations have a start and end statement. Since Soot can optimize
+            // away these statements, we would need to be sure that we update the start and end statements
+            // accordingly.
+            logger.warn(String.format(
+                "Local variable annotation found in %s. Writing out of these annotations is not supported yet.",
+                sm.getSignature()));
+          } else if (t instanceof VisibilityParameterAnnotationTag) {
             VisibilityParameterAnnotationTag vpt = (VisibilityParameterAnnotationTag) t;
             ArrayList<VisibilityAnnotationTag> tags = vpt.getVisibilityAnnotations();
             if (tags != null) {

--- a/src/main/java/soot/FastHierarchy.java
+++ b/src/main/java/soot/FastHierarchy.java
@@ -948,7 +948,7 @@ public class FastHierarchy {
     boolean calleeExist = declaringClass.getMethodUnsafe(subsignature) != null;
     for (SootClass concreteType = baseType; concreteType != null;) {
       SootMethod previouslyResolvedMethod = typeToVtbl.get(concreteType, methodSignature);
-      if (previouslyResolvedMethod != null) {
+      if (previouslyResolvedMethod != null && previouslyResolvedMethod.getDeclaringClass() != null) {
         if (!previouslyResolvedMethod.getDeclaringClass().isInterface()) {
           if (baseType != concreteType) {
             typeToVtbl.put(baseType, methodSignature, previouslyResolvedMethod);

--- a/src/main/java/soot/SootClass.java
+++ b/src/main/java/soot/SootClass.java
@@ -109,7 +109,6 @@ public class SootClass extends AbstractHost {
         }
       };
 
-
   /**
    * Lazy initialized array containing some validators in order to validate the SootClass.
    */
@@ -213,7 +212,7 @@ public class SootClass extends AbstractHost {
     int currentLevel = resolvingLevel();
     if (currentLevel < level) {
       String hint = "\nIf you are extending Soot, try to add the following call before calling soot.Main.main(..):\n"
-          + "Scene.v().addBasicClass(" + getName() + "," + levelToString(level) + ");\n"
+          + "Scene.v().addBasicClass(\"" + getName() + "\"," + levelToString(level) + ");\n"
           + "Otherwise, try whole-program mode (-w).";
       throw new RuntimeException("This operation requires resolving level " + levelToString(level) + " but " + name
           + " is at resolving level " + levelToString(currentLevel) + hint);

--- a/src/main/java/soot/asm/AsmMethodSource.java
+++ b/src/main/java/soot/asm/AsmMethodSource.java
@@ -21,6 +21,7 @@ package soot.asm;
  * <http://www.gnu.org/licenses/lgpl-2.1.html>.
  * #L%
  */
+
 import static org.objectweb.asm.Opcodes.ACONST_NULL;
 import static org.objectweb.asm.Opcodes.ALOAD;
 import static org.objectweb.asm.Opcodes.ANEWARRAY;
@@ -337,7 +338,6 @@ public class AsmMethodSource implements MethodSource {
   private Map<AbstractInsnNode, StackFrame> frames;
   private Multimap<LabelNode, UnitBox> trapHandlers;
   private JimpleBody body;
-  private int lastLineNumber = -1;
   private Table<AbstractInsnNode, AbstractInsnNode, Edge> edges;
   private ArrayDeque<Edge> conversionWorklist;
   private Map<AbstractInsnNode, Integer> lineNumberMap;
@@ -542,9 +542,11 @@ public class AsmMethodSource implements MethodSource {
   }
 
   protected void setUnit(AbstractInsnNode insn, Unit u) {
-    int ln = lineNumberMap.getOrDefault(insn, lastLineNumber);
-    if (ln >= 0) {
-      setLineNumber(u, ln);
+    if (lineNumberMap != null) {
+      int ln = lineNumberMap.get(insn);
+      if (ln >= 0) {
+        setLineNumber(u, ln);
+      }
     }
 
     Unit o = units.put(insn, u);
@@ -554,14 +556,12 @@ public class AsmMethodSource implements MethodSource {
   }
 
   protected void setLineNumber(Unit u, int lineNumber) {
-    if (Options.v().keep_line_number()) {
-      Tag lineTag = u.getTag(LineNumberTag.NAME);
-      if (lineTag == null) {
-        lineTag = new LineNumberTag(lineNumber);
-        u.addTag(lineTag);
-      } else if (((LineNumberTag) lineTag).getLineNumber() != lineNumber) {
-        throw new RuntimeException("Line tag mismatch");
-      }
+    Tag lineTag = u.getTag(LineNumberTag.NAME);
+    if (lineTag == null) {
+      lineTag = new LineNumberTag(lineNumber);
+      u.addTag(lineTag);
+    } else if (((LineNumberTag) lineTag).getLineNumber() != lineNumber) {
+      throw new RuntimeException("Line tag mismatch");
     }
   }
 
@@ -1878,10 +1878,6 @@ public class AsmMethodSource implements MethodSource {
     push(opr);
   }
 
-  private void convertLine(LineNumberNode ln) {
-    lastLineNumber = ln.line;
-  }
-
   private void addEdges(AbstractInsnNode cur, AbstractInsnNode tgt1, List<LabelNode> tgts) {
     int lastIdx = tgts == null ? -1 : tgts.size() - 1;
     Operand[] stackss = stack.toArray(new Operand[stack.size()]);
@@ -1891,9 +1887,7 @@ public class AsmMethodSource implements MethodSource {
     tgt_loop: do {
       Edge edge = edges.get(cur, tgt);
       if (edge == null) {
-        // make sure to store last line number to stay sound if the branch that comes later in
-        // bytecode is processed first
-        edge = new Edge(tgt, lastLineNumber);
+        edge = new Edge(tgt);
         edge.prevStacks.add(stackssL);
         edges.put(cur, tgt, edge);
         conversionWorklist.add(edge);
@@ -1972,17 +1966,14 @@ public class AsmMethodSource implements MethodSource {
     worklist.add(new Edge(instructions.getFirst(), new ArrayList<Operand>()));
     conversionWorklist = worklist;
     edges = HashBasedTable.create(instructions.size(), 1);
-    setLineNumberMap();
+    if (Options.v().keep_line_number()) {
+      setLineNumberMap();
+    }
 
     do {
       Edge edge = worklist.pollLast();
       AbstractInsnNode insn = edge.insn;
       stack = edge.stack;
-      // restore line. this is important since we might have traversed the edge that leads to
-      // bytecode far away from the branch statement first and are now processing the statement
-      // right after the branch which should start with the lastLineNumber as it was for the branch
-      // statement
-      lastLineNumber = edge.lastLineNumber == -1 ? lastLineNumber : edge.lastLineNumber;
       edge.stack = null;
       insnLoop: do {
         int type = insn.getType();
@@ -2055,7 +2046,7 @@ public class AsmMethodSource implements MethodSource {
             convertLabel((LabelNode) insn);
             continue;
           case LINE:
-            convertLine((LineNumberNode) insn);
+            // was already handled in setLineNumberMap()
             continue;
           case FRAME:
             // we can ignore it
@@ -2070,6 +2061,7 @@ public class AsmMethodSource implements MethodSource {
   }
 
   private void setLineNumberMap() {
+    lineNumberMap = new HashMap<>(instructions.size() * 2 + 1);
     AbstractInsnNode current = instructions.getFirst();
 
     int lastNumber = -1;
@@ -2317,7 +2309,6 @@ public class AsmMethodSource implements MethodSource {
     labels = LinkedListMultimap.create(4);
     units = new LinkedHashMap<AbstractInsnNode, Unit>(nrInsn);
     frames = new LinkedHashMap<AbstractInsnNode, StackFrame>(nrInsn);
-    lineNumberMap = new HashMap<>(nrInsn);
     trapHandlers = LinkedListMultimap.create(tryCatchBlocks.size());
     body = jb;
     /* retrieve all trap handlers */
@@ -2537,7 +2528,6 @@ public class AsmMethodSource implements MethodSource {
     final AbstractInsnNode insn;
     /* previous stacks at edge */
     final Set<List<Operand>> prevStacks;
-    private int lastLineNumber = -1;
     /* current stack at edge */
     ArrayList<Operand> stack;
 
@@ -2547,9 +2537,8 @@ public class AsmMethodSource implements MethodSource {
       this.stack = stack;
     }
 
-    Edge(AbstractInsnNode insn, int lastLineNumber) {
+    Edge(AbstractInsnNode insn) {
       this(insn, new ArrayList<Operand>(AsmMethodSource.this.stack));
-      this.lastLineNumber = lastLineNumber;
     }
   }
 }

--- a/src/main/java/soot/asm/AsmMethodSource.java
+++ b/src/main/java/soot/asm/AsmMethodSource.java
@@ -542,8 +542,8 @@ public class AsmMethodSource implements MethodSource {
 
   protected void setUnit(AbstractInsnNode insn, Unit u) {
     if (lineNumberMap != null) {
-      int ln = lineNumberMap.get(insn);
-      if (ln >= 0) {
+      Integer ln = lineNumberMap.get(insn);
+      if (ln != null && ln >= 0) {
         setLineNumber(u, ln);
       }
     }

--- a/src/main/java/soot/asm/AsmMethodSource.java
+++ b/src/main/java/soot/asm/AsmMethodSource.java
@@ -1913,7 +1913,7 @@ public class AsmMethodSource implements MethodSource {
         continue tgt_loop;
       }
       edge.stack = new ArrayList<Operand>(stack);
-      conversionWorklist.addFirst(edge);
+      conversionWorklist.addLast(edge);
     } while (i <= lastIdx && (tgt = tgts.get(i++)) != null);
   }
 

--- a/src/main/java/soot/asm/AsmMethodSource.java
+++ b/src/main/java/soot/asm/AsmMethodSource.java
@@ -21,7 +21,6 @@ package soot.asm;
  * <http://www.gnu.org/licenses/lgpl-2.1.html>.
  * #L%
  */
-
 import static org.objectweb.asm.Opcodes.ACONST_NULL;
 import static org.objectweb.asm.Opcodes.ALOAD;
 import static org.objectweb.asm.Opcodes.ANEWARRAY;
@@ -174,6 +173,7 @@ import static org.objectweb.asm.tree.AbstractInsnNode.MULTIANEWARRAY_INSN;
 import static org.objectweb.asm.tree.AbstractInsnNode.TABLESWITCH_INSN;
 import static org.objectweb.asm.tree.AbstractInsnNode.TYPE_INSN;
 import static org.objectweb.asm.tree.AbstractInsnNode.VAR_INSN;
+
 
 import com.google.common.base.Optional;
 import com.google.common.collect.HashBasedTable;
@@ -1914,7 +1914,7 @@ public class AsmMethodSource implements MethodSource {
         continue tgt_loop;
       }
       edge.stack = new ArrayList<Operand>(stack);
-      conversionWorklist.add(edge);
+      conversionWorklist.addFirst(edge);
     } while (i <= lastIdx && (tgt = tgts.get(i++)) != null);
   }
 
@@ -2539,6 +2539,18 @@ public class AsmMethodSource implements MethodSource {
 
     Edge(AbstractInsnNode insn) {
       this(insn, new ArrayList<Operand>(AsmMethodSource.this.stack));
+    }
+
+    @Override
+    public String toString() {
+      StringBuilder sb = new StringBuilder(insn.toString());
+      if (stack != null) {
+        sb.append("\nCurrent stack:");
+        for (Operand i : stack) {
+          sb.append("\n").append(i.toString());
+        }
+      }
+      return sb.toString();
     }
   }
 }

--- a/src/main/java/soot/jimple/toolkits/typing/fast/BytecodeHierarchy.java
+++ b/src/main/java/soot/jimple/toolkits/typing/fast/BytecodeHierarchy.java
@@ -227,7 +227,16 @@ public class BytecodeHierarchy implements IHierarchy {
     } else if (ancestor instanceof NullType) {
       return false;
     } else {
+      makeSureTypeIsLoaded(ancestor);
+      makeSureTypeIsLoaded(child);
       return Scene.v().getOrMakeFastHierarchy().canStoreType(child, ancestor);
+    }
+  }
+
+  private static void makeSureTypeIsLoaded(Type t) {
+    if (t instanceof RefType) {
+      RefType rt = (RefType) t;
+      Scene.v().forceResolve(rt.getClassName(), SootClass.HIERARCHY);
     }
   }
 

--- a/src/main/java/soot/toolkits/scalar/LocalSplitter.java
+++ b/src/main/java/soot/toolkits/scalar/LocalSplitter.java
@@ -44,7 +44,9 @@ import soot.Unit;
 import soot.Value;
 import soot.ValueBox;
 import soot.options.Options;
+import soot.toolkits.exceptions.PedanticThrowAnalysis;
 import soot.toolkits.exceptions.ThrowAnalysis;
+import soot.toolkits.exceptions.UnitThrowAnalysis;
 import soot.toolkits.graph.ExceptionalUnitGraph;
 import soot.toolkits.graph.ExceptionalUnitGraphFactory;
 import soot.util.LocalBitSetPacker;
@@ -96,6 +98,67 @@ public class LocalSplitter extends BodyTransformer {
 
     if (throwAnalysis == null) {
       throwAnalysis = Scene.v().getDefaultThrowAnalysis();
+      if (throwAnalysis instanceof UnitThrowAnalysis
+          && o.throw_analysis() == Options.throw_analysis_auto_select) {
+        /*
+         * Sadly, the JVM is not smart
+         See https://github.com/soot-oss/soot/issues/1951
+         
+         With the standard UnitThrowAnalysis, soot knows that the definition at (1) is only used at (2), since
+         the event handler at (3) cannot be reached before overwriting a at (2).
+         Thus, the local splitter splits the variable and 
+         uses the split variable at (1) and at the right side of (2), but not at (3).
+         
+         Input: 
+    public static void m1(int)
+    {
+        unknown a, $stack2, $stack3, $stack4, A, $stack5, a#1;
+
+        a := @parameter0: int; (1)
+
+     label1:
+        a = a * 2; (2)
+        ...
+      
+     label2:
+        $stack4 := @caughtexception;
+
+        virtualinvoke $stack5.<java.io.PrintStream: void println(int)>(a); (3)
+
+        return;
+
+        catch java.io.IOException from label1 to label2 with label2;
+    }
+    
+    
+    Output with UnitThrowAnalysis:
+        
+    public static void m1(int)
+    {
+        unknown a, $stack2, $stack3, $stack4, A, $stack5, a#1;
+
+        a#1 := @parameter0: int;
+
+     label1:
+        a = a#1 * 2;
+        ...
+      
+     label2:
+        $stack4 := @caughtexception;
+
+        virtualinvoke $stack5.<java.io.PrintStream: void println(int)>(a);
+
+        return;
+
+        catch java.io.IOException from label1 to label2 with label2;
+    }
+    
+    Now, the JVM assumes conservatively that the exception handler could be reached
+    prior to the definition of a. Thus, we use the pedantic throw analysis which is conservative. 
+         */
+        
+        throwAnalysis = PedanticThrowAnalysis.v();
+      }
     }
 
     if (!omitExceptingUnitEdges) {

